xds/internal/balancer/cdsbalancer/clusterHandler.go

```diff
@@ -0,0 +1,189 @@
+package cdsbalancer
+import (
+	"errors"
+	"strings"
+	"sync"
+
+	xdsclient "google.golang.org/grpc/xds/internal/client"
+)
+
+
+// ClusterHandler will be given a name representing a cluster. It will then update the CDS policy
+// constantly with a list of Clusters to pass down to XdsClusterResolverLoadBalancingPolicyConfig
+// in a stream like fashion.
+type ClusterHandler struct {
+	root *ClusterNode
+
+	// A way to get pinged about any updates or errors to a Node in the tree.
+	updateChannel chan string
+
+	// A mutex to protect entire tree.
+	clusterMutex sync.Mutex
+	// A channel to communicate with the cds balancer about any updates.
+	cdsUpdate chan<- []xdsclient.ClusterUpdate
+	// A channel to communicate with the cds balancer about any errors.
+	errChan chan error
+
+	xdsClient      xdsClientInterface
+}
+
+func (ch *ClusterHandler) updateRootCluster(clusterName string) /*Return value - the expectation for a stream back to the CDS Policy*/ {
+	if ch.root == nil {
+		// Construct a root node on first update.
+		ch.root = CreateClusterNode(clusterName, ch.clusterMutex, ch.updateChannel, ch.xdsClient)
+		return
+	}
+	// Check if root cluster was changed. If it was, delete old one and start new one, if not
+	// do nothing.
+	if clusterName != ch.root.clusterUpdate.ServiceName {
+		ch.root.delete()
+		ch.root = CreateClusterNode(clusterName, ch.clusterMutex, ch.updateChannel, ch.xdsClient)
+	}
+}
+
+// Watches for updates on any ClusterNode comprising the tree, and sends out a response to cds balancer.
+// Will get run from cds.
+func (ch *ClusterHandler) watcher() {
```

Comment:
It looks like you want to have a `go watcher()` goroutine to take care of all the updates, but you also made a global mutex.
Usually you only need one of these.

If you have one goroutine to take care of all the updates, all the actions will be made from the same goroutine, the mutex is unnecessary.

Similar for the mutex, if all actions are protected by the mutex, even if they are called from different goroutines, they are sync'ed.

Reply:
And I think callback+global mutex is easier in this case. And this is what I would define.

```go
func watchCluster(clusterName string, client xdsClientInterface, callback func([]xdsclient.ClusterUpdate, error)) func() {
	ch := ClusterHandler{
		xdsClient: client,
		callback:  callback,
	}
	ch.root, ch.cancel = newNode(clusterName, client, ch)
	return ch.close
}

func (h ClusterHandler) constructClusterUpdate() {
	update, err := root.constructClusterUpdate()
	if err == nil {
		h.callback(update, nil)
	}
	// ...
}

func newNode(clusterName string, client xdsClientInterface, topLevelHandler *ClusterHandler) (*ClusterNode, func()) {
	n := &ClusterNode{}
	n.cancel = client.WatchCluster(clusterName, func(update xdsclient.ClusterUpdate, err error) {
		topLevelHandler.clusterMutex.Lock()
		defer topLevelHandler.clusterMutex.Unlock()
		n.handleResp(update, err)
		topLevelHandler.constructClusterUpdate()
	})
}
```

Reply:
In reference to your first comment, the global mutex isn't used for synchronizing updates. Logically, I view the clusterMutex as protecting **the whole tree** rather than **the update events**. The reason you need to protect the construction of an update to **the whole tree** is that a root node could get a delete as you are building the update, now you have undefined behavior/concurrent memory accesses.

Reply:
Thus, on the construction of the update from the root node, I call the mutex, logically synchronizing accesses to the **whole tree's memory**, which constructs the whole list every time a child node pings it that it received an update. Obviously, the initial pings before the full tree is constructed will have to be ignored, but we discussed this and I took care of this scenario in errors from the recursive construct cluster update, which the handler just ignores.

Reply:
Any **actual errors** (received from the CDS Watch, not an error of not having received a first update which will be ignored), for a child cluster will have to be sent down to xds_cluster_resolver.

Reply:
I thought about this a lot. You'll see the style I had (it's logically the same as what you wrote, but in the style I prefer. You'll see when I assign back to you and let me know what you think).